ceno_zkvm/src/scheme/cpu/mod.rs

```diff
@@ -583,6 +517,124 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> TowerProver<CpuBacke
 impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> MainSumcheckProver<CpuBackend<E, PCS>>
     for CpuProver<CpuBackend<E, PCS>>
 {
+    #[allow(clippy::type_complexity)]
+    #[tracing::instrument(
+        skip_all,
+        name = "build_main_witness",
+        fields(profiling_3),
+        level = "trace"
+    )]
+    fn build_main_witness<'a, 'b>(
+        &self,
+        composed_cs: &ComposedConstrainSystem<E>,
+        input: &ProofInput<'a, CpuBackend<E, PCS>>,
+        challenges: &[E; 2],
+    ) -> (Vec<ArcMultilinearExtension<'b, E>>, bool)
+    where
+        'a: 'b,
+    {
+        let ComposedConstrainSystem {
+            zkvm_v1_css: cs,
+            gkr_circuit,
+        } = composed_cs;
+        let num_instances = input.num_instances;
+        let log2_num_instances = input.log2_num_instances();
+        let num_var_with_rotation = log2_num_instances + composed_cs.rotation_vars().unwrap_or(0);
+        let num_instances_with_rotation = num_instances << composed_cs.rotation_vars().unwrap_or(0);
+
+        // sanity check
+        assert_eq!(input.witness.len(), cs.num_witin as usize);
+
+        // structural witness can be empty. In this case they are `eq`, and will be filled later
+        assert!(
+            input.structural_witness.len() == cs.num_structural_witin as usize
+                || input.structural_witness.is_empty(),
+        );
+        assert_eq!(input.fixed.len(), cs.num_fixed);
+
+        // check all witness size are power of 2
+        assert!(
+            input
+                .witness
+                .iter()
+                .all(|v| { v.evaluations().len() == 1 << num_var_with_rotation })
+        );
+
+        if let Some(gkr_circuit) = gkr_circuit {
+            // opcode must have at least one read/write/lookup
+            assert!(
+                cs.lk_expressions.is_empty()
+                    || !cs.r_expressions.is_empty()
+                    || !cs.w_expressions.is_empty(),
+                "assert opcode circuit"
+            );
+
+            let (_, gkr_circuit_out) = Self::gkr_witness(
+                gkr_circuit,
+                num_instances_with_rotation,
+                &input.witness,
+                &input.fixed,
+                &input.public_input,
+                challenges,
+            );
+            (gkr_circuit_out.0.0, true)
+        } else {
```

Comment:
`else` condition only needed for table circuit, because table circuit do not need main-sumcheck, thus I didn't refactor it into gkr_iop framework at this moment